graphql_compiler/query_formatting/common.py

```diff
@@ -74,49 +56,55 @@ def _deserialize_anonymous_json_argument(expected_type: GraphQLScalarType, value
+def _validate_json_scalar_argument(name: str, expected_type: GraphQLScalarType, value: Any) -> None:
+    """Validate that the JSON serialized scalar argument has the expected python type."""
+    expected_python_types = _ALLOWED_JSON_SCALAR_TYPES.get(expected_type.name)
+    if expected_python_types is None:
+        raise AssertionError(
+            f"Got unsupported GraphQL type {expected_type} for argument {name} with value {value}."
         )
+    elif (
+        # We explicitly disallow passing boolean values for non-boolean types
+        (isinstance(value, bool) and not is_same_type(GraphQLBoolean, expected_type))
+        or not isinstance(value, expected_python_types)
+    ):
+        _raise_invalid_type_error(name, expected_python_types, value)
+
+
+_CUSTOM_SCALAR_DESERIALIZATION_FUNCTIONS = MappingProxyType(
+    {GraphQLInt.name: int, GraphQLFloat.name: float}
+)
+_SCALAR_DESERIALIZATION_FUNCTIONS = MappingProxyType(
+    {
+        name: _CUSTOM_SCALAR_DESERIALIZATION_FUNCTIONS.get(name, graphql_type.parse_value)
+        for name, graphql_type in SCALAR_TYPE_NAME_TO_VALUE.items()
+    }
+)
 
-    name_to_custom_type = {graphql_type.name: graphql_type for graphql_type in CUSTOM_SCALAR_TYPES}
-    # Parse the value. In most cases we can use the default GraphQL parser, but there are some
-    # special cases where we are more permissive than the default.
-    # By default, strings cannot be parsed to float.
-    if is_same_type(expected_type, GraphQLFloat):
-        return float(value)
-    # Use the default GraphQL parser to parse the value
-    elif expected_type.name in name_to_custom_type:
-        # Since we cannot serialize the parse_value function of custom scalar types when
-        # serializing a schema, it is possible that the parse_value is incorrectly set. We must,
-        # therefore, use the parse_value function of the original scalar type definition.
-        return name_to_custom_type[expected_type.name].parse_value(value)
-    else:
-        return expected_type.parse_value(value)
+
+def _deserialize_json_scalar_argument(name, expected_type: GraphQLScalarType, value: Any) -> Any:
+    """Deserialize the json serialized scalar argument."""
+    _validate_json_scalar_argument(name, expected_type, value)
+    try:
+        return _SCALAR_DESERIALIZATION_FUNCTIONS[expected_type.name](value)
+    except (ValueError, TypeError) as e:
+        raise GraphQLInvalidArgumentError("Error parsing argument {}: {}".format(name, e))
 
 
 def deserialize_json_argument(
     name: str,
-    expected_type: Union[GraphQLNonNull[GraphQLScalarType], GraphQLScalarType],
+    expected_type: Union[
+        GraphQLScalarType,
+        GraphQLList[GraphQLScalarType],
+        GraphQLList[GraphQLNonNull[GraphQLScalarType]],
+        GraphQLNonNull[GraphQLScalarType],
+        GraphQLNonNull[GraphQLList[GraphQLScalarType]],
+        GraphQLNonNull[GraphQLList[GraphQLNonNull[GraphQLScalarType]]],
```

Comment:
Thought of making this type annotation a variable but got a mypy error that pointed me to the following link:
https://mypy.readthedocs.io/en/latest/common_issues.html#variables-vs-type-aliases

Reply:
I got the equivalent of the "error: Variable "__main__.tp" is not valid as a type" error you can see in the link.

Reply:
You don't want to make it a variable, you want to make it a type alias — that's precisely the distinction that `mypy` attempted to teach you at that link. A type alias will do what you want here, so please give the link one more shot and then let's discuss if it's still confusing.

Reply:
I am not sure about what I was doing wrong earlier, but I was able to fix this.